pkg/storj/metainfo.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storj
+
+import (
+	"context"
+)
+
+// Metainfo represents a database for storing meta-info about objects
+type Metainfo interface {
+	// MetainfoLimits returns limits for this metainfo database
+	Limits() (MetainfoLimits, error)
+
+	// CreateBucket creates a new bucket with the specified information
+	// Database automatically sets different values in the information
+	CreateBucket(ctx context.Context, bucket string, info *Bucket) (Bucket, error)
+	// DeleteBucket deletes bucket
+	DeleteBucket(ctx context.Context, bucket string) error
+	// GetBucket gets bucket information
+	GetBucket(ctx context.Context, bucket string) (Bucket, error)
+
+	// GetObject returns information about an object
+	GetObject(ctx context.Context, bucket string, path Path) (Object, error)
+	// GetObjectStream returns interface for reading the object stream
+	GetObjectStream(ctx context.Context, bucket string, path Path) (ReadOnlyStream, error)
+
+	// CreateObject creates an uploading object and returns an interface for uploading Object information
+	CreateObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
+	// ModifyObject creates an interface for modifying an existing object
+	ModifyObject(ctx context.Context, bucket string, path Path, info Object) (MutableObject, error)
+	// DeleteObject deletes an object from database
+	DeleteObject(ctx context.Context, bucket string, path Path) error
+
```

Comment:
How about adding CopyObject and MoveObject methods too?

Reply:
I'm not yet sure whether this is the right level for implementing this. For `MoveObject` probably since there we don't have to communicate with the existing nodes. For `CopyObject`, if we need to redownload and decrypt/encrypt then it's probably the wrong level.

Reply:
Ah, right... this would become an method on `Uplink` rather than `Metainfo`, since it involves communication and coordination with the storage node.

We can add other methods to make this possible at the metainfo level. But either way we might have:
1. uplink.CopyObject,
2. which calls metainfo.CreateObject,
3. starts downloading the segment info,
4. gets appropriate allocation for copy from satellite,
5. sends a Copy request to storage nodes,
6. calls AddSegment with the information.